jobs/mysql_backup_job.go

```diff
@@ -0,0 +1,190 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jobs
+
+import (
+	"bytes"
+	"context"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	xtrabackupBin = "xtrabackup"
+	xbcloudBin    = "xbcloud"
+)
+
+// MySQLBackupJob implements Job for MySQL backup.
+type MySQLBackupJob struct {
+	id       string
+	timeout  time.Duration
+	l        *logrus.Entry
+	name     string
+	dsn      string
+	location BackupLocationConfig
+}
+
+// S3LocationConfig contains required properties for accessing S3 Bucket.
+type S3LocationConfig struct {
+	Endpoint     string
+	AccessKey    string
+	SecretKey    string
+	BucketName   string
+	BucketRegion string
+}
+
+// BackupLocationConfig groups all backup locations configs.
+type BackupLocationConfig struct {
+	S3Config *S3LocationConfig
+}
+
+// NewMySQLBackupJob constructs new Job for MySQL backup.
+func NewMySQLBackupJob(id string, timeout time.Duration, name, dsn string, locationConfig BackupLocationConfig) *MySQLBackupJob {
+	return &MySQLBackupJob{
+		id:       id,
+		timeout:  timeout,
+		l:        logrus.WithFields(logrus.Fields{"id": id, "type": "mysql_backup", "name": name}),
+		name:     name,
+		dsn:      dsn,
+		location: locationConfig,
+	}
+}
+
+// ID returns job id.
+func (j *MySQLBackupJob) ID() string {
+	return j.id
+}
+
+// Type returns job type.
+func (j *MySQLBackupJob) Type() string {
+	return "mysql_backup"
+}
+
+// Timeout returns job timeout.
+func (j *MySQLBackupJob) Timeout() time.Duration {
+	return j.timeout
+}
+
+func (j *MySQLBackupJob) Run(ctx context.Context, send Send) error {
+	t := time.Now()
+	j.l.Info("MySQL backup started")
+
+	mysqlConfig, err := mysql.ParseDSN(j.dsn)
```

Comment:
I would suggest to pass exact values (`host`, `port` and e.t.c.) from backend instead of passing dsn and parsing it again